brayns/io/MolecularSystemReader.cpp

```diff
@@ -0,0 +1,220 @@
+#include <brayns/io/MeshLoader.h>
+#include <fstream>
+
+namespace brayns
+{
+
+MolecularSystemReader::MolecularSystemReader(
+    const GeometryParameters& geometryParameters )
+    : _geometryParameters( geometryParameters )
+{
+}
+
+bool MolecularSystemReader::import( Scene& scene )
+{
+    const auto& configuration = _geometryParameters.getMolecularSystemConfig();
+    // Load molecular system configuration
+    std::ifstream configurationFile( configuration, std::ios::in );
+    if( !configurationFile.good( ))
+    {
+        BRAYNS_ERROR << "Could not open file " << configuration << std::endl;
+        return false;
+    }
+
+    std::map< std::string, std::string > parameters;
+    std::string line;
+    while( std::getline( configurationFile, line ))
+    {
+        std::stringstream lineStream( line );
+        std::string key, value;
+        lineStream >> key >> value;
+        parameters[ key ] = value;
+    }
+    configurationFile.close();
+
+    const auto& proteinFolder = parameters["ProteinFolder"];
+    const auto& meshFolder = parameters["MeshFolder"];
+    const auto& descriptor = parameters["SystemDescriptor"];
+    const auto& positions = parameters["ProteinPositions"];
+
+    BRAYNS_INFO << "Loading biological assembly" << std::endl;
+    BRAYNS_INFO << "Protein folder    : " << proteinFolder << std::endl;
+    BRAYNS_INFO << "Mesh folder       : " << meshFolder << std::endl;
+    BRAYNS_INFO << "System descriptor : " << descriptor << std::endl;
+    BRAYNS_INFO << "Protein positions : " << positions << std::endl;
+
+    std::ifstream descriptorFile( descriptor, std::ios::in );
+    if( !descriptorFile.good( ))
+    {
+        BRAYNS_ERROR << "Could not open file " << descriptor << std::endl;
+        return false;
+    }
+
+    // Load list of proteins
+    std::map< size_t, std::string > proteins;
+    while( std::getline( descriptorFile, line ))
+    {
+        std::stringstream lineStream( line );
+
+        std::string protein;
+        size_t id;
+        size_t instances;
+        lineStream >> protein >> id >> instances;
+        if( protein == "" )
+            continue;
+
+        proteins[id] = protein;
+
+        if( proteinFolder == "" )
+            continue;
+
+        const std::string pdbFilename( proteinFolder + '/' + protein + ".pdb" );
+        std::ifstream pdbFile( pdbFilename, std::ios::in );
+        if( pdbFile.good( ))
+            pdbFile.close();
+        else
+        {
+            // PDB file not present in folder, download it from RCSB.org
+            std::string command;
+            command = "wget http://www.rcsb.org/pdb/files/";
+            command += protein;
+            command += ".pdb";
+            command += " -P ";
+            command += proteinFolder;
+            int status = system( command.c_str( ));
+            BRAYNS_INFO << command << ": " << status << std::endl;
+        }
+    }
+    descriptorFile.close();
+
+    // Load proteins according to specified positions
```

Comment:
for readability you could move the comment-marked blocks to private methods. Would also 'solve' the file.close() at the end of each block.